salt/modules/ansiblegate.py

```diff
@@ -0,0 +1,250 @@
+
+        :param module:
+        :return:
+        '''
+        m_ref = self._modules_map.get(module)
+        if m_ref is None:
+            raise LoaderError('Module "{0}" was not found'.format(module))
+        mod = importlib.import_module('ansible.modules{0}'.format(
+            '.'.join([elm.split('.')[0] for elm in m_ref.split(os.path.sep)])))
+
+        return mod
+
+    def get_modules_list(self, pattern=None):
+        '''
+        Return module map references.
+        :return:
+        '''
+        if pattern and '*' not in pattern:
+            pattern = '*{0}*'.format(pattern)
+        modules = []
+        for m_name, m_path in self._modules_map.items():
+            m_path = m_path.split('.')[0]
+            m_name = '.'.join([elm for elm in m_path.split(os.path.sep) if elm])
+            if pattern and fnmatch.fnmatch(m_name, pattern) or not pattern:
+                modules.append(m_name)
+        return sorted(modules)
+
+    def resolve(self):
+        log.debug('Resolving Ansible modules')
+        self._modules_map = self._get_modules_map()
+        return self
+
+    def install(self):
+        log.debug('Installing Ansible modules')
+        return self
+
+
+class AnsibleModuleCaller(object):
+    def __init__(self, resolver):
+        self._resolver = resolver
+
+    def call(self, module, *args, **kwargs):
+        '''
+        Call an Ansible module by invoking it.
+        :param module: the name of the module.
+        :param args: Arguments to the module
+        :param kwargs: keywords to the module
+        :return:
+        '''
+
+        module = self._resolver.load_module(module)
+        if not hasattr(module, 'main'):
+            raise CommandExecutionError('This module is not callable '
+                                        '(see "ansible.help {0}")'.format(module.__name__.replace('ansible.modules.',
+                                                                                                  '')))
+        if args:
+            kwargs['_raw_params'] = ' '.join(args)
+        js_args = '{{"ANSIBLE_MODULE_ARGS": {args}}}'.format(args=json.dumps(kwargs))
+        js_out = subprocess.Popen(["echo", "{0}".format(js_args)], stdout=subprocess.PIPE)
+        md_exc = subprocess.Popen(['python', module.__file__],
+                                  stdin=js_out.stdout, stdout=subprocess.PIPE)
+        js_out.stdout.close()
+        js_out, js_err = md_exc.communicate()
+
+        try:
+            out = json.loads(js_out)
+        except ValueError as ex:
+            out = {'Error': (js_err and (js_err + '.') or str(ex))}
+            if js_out:
+                out['Given JSON output'] = js_out
+            return out
+
+        if 'invocation' in out:
+            del out['invocation']
+
+        return out
+
+
+_resolver = None
+_caller = None
+
+
+def _set_callables(modules):
+    '''
+    Set all Ansible modules callables
+    :return:
+    '''
+    def _mkf(cmd_name, doc):
```

Comment:
A more descriptive function name here would help make this more maintainable in the future. :]

Reply:
OK